### PR TITLE
Per tenant config + pitr

### PR DESCRIPTION
**Pull request by lubennikovaav**
New version of @knizhnik PR #1332 

This PR implements per-tenant config and also adds new `pitr_interval` parameter.
Maybe it's worth to split pitr part into a separate commit, as there were some questions about the proper implementation of pitr.

TenantConf is still stored as a part of the global PageserverConf, but one can also pass per-tenant conf parameters when tenant is created or using newAPI or  `zenith tenant config` CLI command.
This custom TenantConfOpt is stored in LayeredRepository struct.  
To get the parameter value, check TenantConfOpt, if it is not set, use value from global config.

When updated, tenant conf is saved to the tenant's directory, to recover it on pageserver restart.

To support tenant migraiton, this config must be stored in the console along with other tenant information. @ololobus, FYI

### CODE REVIEW

**Comment by ololobus** [comment]
@alubennikova does pageserver have an API endpoint to get / put this per-tenant config? Without it it will will be impossible to mange it from the console, we don't use ssh anymore

Also I wonder should the PITR config be per-timeline instead? It seems to be dependent on amount of WAL we keep and it's stored per-timeline, isn't it?

**Comment by knizhnik** [comment]
> @alubennikova does pageserver have an API endpoint to get / put this per-tenant config? Without it it will be impossible to mange it from the console, we don't use ssh anymore
> 

Right now tenant configuration is specified by console when tenant is created.
It is possible to add API to change it on the fly but it has not yet be done.
Also in this  case we have to propagate this changes to all active threads (GC, compactor,...)

> Also I wonder should the PITR config be per-timeline instead? It seems to be dependent on amount of WAL we keep and it's stored per-timeline, isn't it?

No, it is stored per tenant.
IMHO there is no sense to have it per timeline, but I may be wrong.
Why do we need per-tenant config?
Because we expect that we will have have some VIP customers for we we can adjust page server parameters.
But I do not see reasons to do it per timeline.
Also please notice that GC and compactor threads are handling all tenants timeline.

**Comment by lubennikovaav** [comment]
> > @alubennikova does pageserver have an API endpoint to get / put this per-tenant config? Without it it will be impossible to mange it from the console, we don't use ssh anymore
> 
> Right now tenant configuration is specified by console when tenant is created. It is possible to add API to change it on the fly but it has not yet be done.

@ololobus 
I've implemented put API: `put("/v1/tenant", tenant_config_handler)`

Now I think that is probably should be
`put("/v1/tenant/:tenant_id/config", tenant_config_handler)
get("/v1/tenant/:tenant_id/config", tenant_config_handler)`

Please, review `openapi_spec.yml` and `routes.rs` changes

>  Also in this case we have to propagate this changes to all active threads (GC, compactor,...)
@knizhnik 
timeline's TenantConf is stored as an Arc to the repo's TenantConf. So when it is changed, all threads should see the change.
If I wrote it right, it should work) Please review the code.

**Review by ololobus — COMMENTED** [review]
> Now I think that is probably should be
>
> put("/v1/tenant/:tenant_id/config", tenant_config_handler)
> get("/v1/tenant/:tenant_id/config", tenant_config_get_handler)

I think that how you did is fine. IMO, it's a bit inconvenient to pass key (tenant_id) inside body instead of path argument, but I don't see any problems with it.

About get, I don't think we really need it now. It seems that we only need to pass some default config on creation and maybe update it later. So post + put should be enough.

**Review by knizhnik — COMMENTED** [review]
Originally TenantConf was copied and requires no synchronization on access.
Now it is changed to Arc<RwLock<...>>.
It allows to update configuration online, without restarting threads.
But there are some drawbacks:
1. Synchronization overhead
2. If parameter value is storedin some local variable, then changing configuration will nottake effect
3. Possibility of hangout and even deadlocks.

So may be online update of configuration is really important.
But in this case this three items has to be carefully checked.
I have not noticed places where config is accessed so frequently that is can become a bottleneck. Also I have not noticed placed where config value is cached in some local variable.
But concerning 3, there are placed like
```
std::thread::sleep(tenant_conf.compaction_period);
```
Please notice that here we sleep with RwLock locked. Ansd any attempt to update confiuration will be blocked on this lock.

**Comment by lubennikovaav** [comment]
> But concerning 3, there are placed like
> 
> ```
> std::thread::sleep(tenant_conf.compaction_period);
> ```
> 
> Please notice that here we sleep with RwLock locked. Ansd any attempt to update confiuration will be blocked on this lock.

I don't think, that we sleep with RwLock, it is released as soon, as we exit `get_tenant_conf()`.
I've tested configuration update and it works fine (set compaction_period=100, update it to some smaller value while the thread sleeps, check the config file - the new value is there).
The change itself will be applied on the next wake of the loop. I think it is expected behavior.

**Comment by lubennikovaav** [comment]
@LizardWizzard, @SomeoneToIgnore 
please, take a look at this PR. I need some help to connect it with tenant relocation.
AFAIK, you are refactoring it right now, so I disabled tenant config loading in the latest commit to avoid `test_tenant_relocation` failures. It means that tenant-specific config now doesn't survive relocation. 

I'd like to merge this PR anyway and fix it later, when we will have full support of the tenant relocation elsewhere.
What do you think?

File: pageserver/src/http/openapi_spec.yml
Comment (lubennikovaav) [reply pageserver/src/http/openapi_spec.yml]: Thank you, it was a copy-paste mistake. Fixed

Comment (ololobus) [reply pageserver/src/http/openapi_spec.yml]: Also we should put this updated spec into https://github.com/neondatabase/cloud/blob/main/spec/pageserver.yml, run `make generate-all` and open a PR to check that everything works well after all these changes

**Comment by knizhnik** [comment]
> 

I see, it is copied...
Sorry for false alarm.

**Review by knizhnik — APPROVED** [review]

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [inline pageserver/src/layered_repository.rs:643]: nit: probably `tenantconf_path` can be just `path(...)` for simplicity if called like that

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [inline pageserver/src/layered_repository.rs:2029]: A possible "bloat" issue still persists here. It can arise from modification time change because of layer upload/download. If we are not solving it now let's mention it in a comment with reasoning or create separate issue for a needed fix.

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [inline pageserver/src/layered_repository.rs:2380]: `Vec::new()` -> `vec![]`

File: pageserver/src/page_service.rs
Comment (LizardWizzard) [inline pageserver/src/page_service.rs:676]: Do we really need it? What is the use case? It is already available through http api

File: pageserver/src/layered_repository.rs
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:646]: > 
```suggestion
        let mut conf_content = r#"# This file contains a specific per-tenant's config.
```

File: pageserver/src/tenant_config.rs
Comment (LizardWizzard) [inline pageserver/src/tenant_config.rs:69]: Please use `derive(Default)` or `impl Default for TenantConfOpt`

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [inline pageserver/src/layered_repository.rs:475]: IMO using getters instead of merging configs every time we need them is a bit cleaner. E g in the compaction thread we're using only `compaction_period`

File: pageserver/src/layered_repository.rs
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:1078]: > .unwrap_or

I see that `TenantConfOpt` is used in the layered repository only, and then its values are either used (if present), or defaults are used instead.
That all feels a bit odd, cannot we merge those values once during the init? 

We sure still need the `TenantConfOpt` to parse things from requests/configs, but why do we store it, not the `TenantConf`?

File: pageserver/src/tenant_config.rs
Comment (LizardWizzard) [inline pageserver/src/tenant_config.rs:125]: can be `impl From`. Also I see that it is used only in tests, so maybe makes sense to move it to tests module as a separate helper

File: pageserver/src/tenant_mgr.rs
Comment (LizardWizzard) [inline pageserver/src/tenant_mgr.rs:80]: Lets use a result here instead of a panic. Because this function is used in `apply_timeline_sync_status_updates` so if it panics this will affect other tenants

File: pageserver/src/tenant_config.rs
Comment (LizardWizzard) [inline pageserver/src/tenant_config.rs:58]: nit: naming suggestion: `TenantConfOverrides` this way it'll indicate that here we override global config

File: test_runner/batch_others/test_tenant_conf.py
Comment (SomeoneToIgnore) [inline test_runner/batch_others/test_tenant_conf.py:8]: I propose to test the config update here too.
Would be amazing to update it to something ridiculous and then try to run some operation to see if the config change had influenced that, but getting the value is also ok, I guess.

File: control_plane/src/storage.rs
Comment (SomeoneToIgnore) [inline control_plane/src/storage.rs:346]: Up to you, but I'd rather had some method to parse the strings into `TenantConfOpt` (maybe put the toml parsing into that module too and add some unit tests for both?) and pass it around, not some odd map.

**Review by SomeoneToIgnore — COMMENTED** [review]
I don't have a solid perspective on tenant relocation, so it looks ok to me to do what you've proposed, but maybe @LizardWizzard has better ideas.

Might it be simpler to use a default config for every tenant by default, if nothing provided?

**Review by LizardWizzard — COMMENTED** [review]
+1 for fixing tenant relocation tests separately. I have an idea on how to do that without too much code changes

**Comment by SomeoneToIgnore** [comment]
Also feels like we could use more unit tests on parsing at least.

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [reply pageserver/src/layered_repository.rs:2380]: This is my subjective opinion :) Feel free to leave as is

File: pageserver/src/layered_repository.rs
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:1078]: I agree that it looks strange, but this implementation allows us to preserve information about what tenant-specific settings are set. `persist_tenant_conf()` only saves overridden parameters for the same reason.
This is necessary allow global config updates.
I.e. we want to update `gc_horizon` in PageServerConf for all tenants that do not use custom values.
Not sure, if this scenario will be used very often, because it requires a pageserver restart, but technically it is possible.

File: pageserver/src/config.rs
Comment (lubennikovaav) [reply pageserver/src/config.rs:272]: TBH, I was trying to avoid using this Builder machinery. 
All TenantConf options have default values, so it's totally fine if some of them are absent.

File: pageserver/src/layered_repository.rs
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:625]: I did it this way to unify this code with global config parsing.
Now we have `[tenant_config] ` table in both files.

File: pageserver/src/bin/pageserver.rs
Comment (lubennikovaav) [reply pageserver/src/bin/pageserver.rs:249]: Oh, cool. I didn't know that. I'll leave it as-is if you don't mind.
It's unrelated to the PR anyway, I just noticed a typo in the log.

**Comment by ololobus** [comment]
Pageserver from this branch responds with 500 on create timeline request https://github.com/neondatabase/cloud/runs/6093559286?check_suite_focus=true#step:28:137

I think this is because I asked to rename `new_tenant_id` to just `tenant_id` %) Maybe keep both for backward compatibility for some time? Otherwise, all such breaking changes require to deploy both storage and console, but we are not ready to deploy new storage on prod, IIRC. So we can block releases

File: pageserver/src/config.rs
Comment (LizardWizzard) [reply pageserver/src/config.rs:272]: I'm ok with avoiding builder machinery for tenant conf, but l'm more about the tenant_conf field in pageserver config. Builder already used there so IMO better option is to handle this field the same way as the others

File: pageserver/src/http/routes.rs
Comment (lubennikovaav) [reply pageserver/src/http/routes.rs]: Fixed

File: pageserver/src/config.rs
Comment (lubennikovaav) [reply pageserver/src/config.rs]: I think we need to keep TenantConfOpt, please see the discussion above:
https://github.com/neondatabase/neon/pull/1504#discussion_r854214360

I'll add some comment.

File: pageserver/src/layered_repository.rs
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:475]: Agree. I'll implement getters.
Though, we still need get_tenant_conf() to implement show.

File: pageserver/src/layered_repository.rs
Comment (LizardWizzard) [reply pageserver/src/layered_repository.rs:625]: I'm was not familiar with toml crate api, I thought that types for table entries and plain values should be the same, but unfortunately, it is not true

File: pageserver/src/config.rs
Comment (LizardWizzard) [reply pageserver/src/config.rs]: Got it, makes sense

File: pageserver/src/page_service.rs
Comment (LizardWizzard) [reply pageserver/src/page_service.rs:676]: Hm, for some reason I thought that there is already a `get` endpoint for config. Personally, I'm leaning towards using http api as much as possible without duplicating functionality in postgres protocol. But if people find these helpers useful let's keep them

File: pageserver/src/layered_repository.rs
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:609]: https://github.com/neondatabase/neon/issues/1555

**Comment by lubennikovaav** [comment]
> Pageserver from this branch responds with 500 on create timeline request https://github.com/neondatabase/cloud/runs/6093559286?check_suite_focus=true#step:28:137
> 
> I think this is because I asked to rename `new_tenant_id` to just `tenant_id` %) Maybe keep both for backward compatibility for some time? Otherwise, all such breaking changes require to deploy both storage and console, but we are not ready to deploy new storage on prod, IIRC. So we can block releases

I just left old name in CreateTenant and introduced new endpoint /t1/tenant/config.
There's some code duplication, but it doesn't hurt. We can clean it up later.

File: pageserver/src/http/openapi_spec.yml
Comment (lubennikovaav) [reply pageserver/src/http/openapi_spec.yml]: https://github.com/neondatabase/cloud/pull/1070

**Pull request merged**
